src/core/lib/event_engine/poller.h

```diff
@@ -30,20 +29,22 @@ namespace experimental {
 // Work(...).
 class Poller {
  public:
-  // This initial vector size may need to be tuned
-  using Events = absl::InlinedVector<EventEngine::Closure*, 5>;
-  struct DeadlineExceeded {};
-  struct Kicked {};
-  using WorkResult = absl::variant<Events, DeadlineExceeded, Kicked>;
+  enum class WorkResult { kOk, kDeadlineExceeded, kKicked };
 
   virtual ~Poller() = default;
-  // Poll once for events, returning a collection of Closures to be executed.
+  // Poll once for events and process received events. The callback function
+  // "poll_again" is expected to be schduled asynchronously prior to processing
+  // received events. The callback's responsibility primarily is to call
+  // Poller::Work again. This would ensure that the next polling cycle would run
+  // as quickly as possible to ensure continuous polling.
   //
   // Returns:
-  //  * absl::AbortedError if it was Kicked.
-  //  * absl::DeadlineExceeded if timeout occurred
-  //  * A collection of closures to execute, otherwise
-  virtual WorkResult Work(EventEngine::Duration timeout) = 0;
+  //  * Poller::WorkResult::kKicked if it was Kicked.
+  //  * Poller::WorkResult::kDeadlineExceeded if timeout occurred
+  //  * Poller::WorkResult::kOk, otherwise indicating that the callback function
+  //  was schduled to run asynchonously and some events were processed.
+  virtual WorkResult Work(EventEngine::Duration timeout,
+                          absl::AnyInvocable<void()> poll_again) = 0;
```

Comment:
suggest making this `absl::FunctionRef<void()> schedule_poll_again`
and calling it directly in implementations

right now we'd probably write `Work(... [this] { Run(Poll); });` and that's fine, but it would also allow us to:
1. use `Closure` instead of `AnyInvocable` if that makes sense
2. mutate & see stack variables at the `Work` call site - which might be useful to save an allocation down the track

Reply:
Updated to this model in the latest commit